ntp-proto/src/packet.rs

```diff
@@ -242,6 +242,36 @@ impl NtpHeader {
     pub fn is_kiss_rstr(&self) -> bool {
         self.is_kiss() && self.reference_id.is_rstr()
     }
+
+    pub fn timestamp_response<C: NtpClock>(
+        system: &SystemSnapshot,
+        input: NtpHeader,
+        recv_timestamp: NtpTimestamp,
+        clock: &mut C,
+    ) -> Self {
+        Self {
+            mode: NtpAssociationMode::Server,
+            stratum: system.stratum,
+            origin_timestamp: input.transmit_timestamp,
+            receive_timestamp: recv_timestamp,
+            reference_id: system.reference_id,
+            poll: input.poll,
+            precision: system.precision.log2(),
+            root_delay: system.root_delay,
+            root_dispersion: system.root_dispersion,
+            // Timestamp must be last to make it as accurate as possible.
+            transmit_timestamp: clock.now().expect("Failed to read time"),
+            ..Self::new()
+        }
+    }
+
+    pub fn rate_limit_response() -> Self {
+        Self {
+            stratum: 0, // indicates a kiss code
+            reference_id: ReferenceId::KISS_RATE,
+            ..Self::new()
```

Comment:
Missing the origin timestamp setting here. Note that this is required to be done otherwise the remote will ignore the rate as being not a response to its packet.